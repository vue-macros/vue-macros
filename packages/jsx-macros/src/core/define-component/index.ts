import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import { walkIdentifiers } from '@vue/compiler-sfc'
import { restructure } from '../restructure'
import type { FunctionalNode, RootMapValue } from '..'
import type { OptionsResolved } from '../..'
import { transformAwait } from './await'
import { transformReturn } from './return'
import type { ObjectExpression, Program } from '@babel/types'

export function transformDefineComponent(
  root: FunctionalNode,
  propsName: string,
  map: RootMapValue,
  s: MagicStringAST,
  ast: Program,
  options: OptionsResolved,
): void {
  if (!map.defineComponent) return

  const defineComponentName = s.sliceNode(map.defineComponent.callee)
  if (
    defineComponentName &&
    !['defineComponent', 'defineVaporComponent'].includes(defineComponentName)
  ) {
    importHelperFn(s, 0, 'defineComponent', defineComponentName)
  }

  let hasRestProp = false
  const props: Record<string, string | null> = {}
  if (root.params[0]) {
    if (root.params[0].type === 'Identifier') {
      getWalkedIds(root, propsName).forEach((id) => (props[id] = null))
    } else {
      const restructuredProps = restructure(s, root, {
        generateRestProps: (restPropsName, index, list) => {
          if (index === list.length - 1) {
            hasRestProp = true
            const useAttrs = importHelperFn(s, 0, 'useAttrs')
            return `const ${restPropsName} = ${useAttrs}()`
          }
        },
      })
      for (const prop of restructuredProps) {
        if (prop.path.endsWith('props') && !prop.isRest) {
          props[prop.name] = prop.isRequired ? '{ required: true }' : null
        }
      }
    }
  }

  for (const { expression, isRequired } of map.defineModel || []) {
    const modelOptions =
      expression.arguments[0]?.type === 'ObjectExpression'
        ? expression.arguments[0]
        : expression.arguments[1]?.type === 'ObjectExpression'
          ? expression.arguments[1]
          : undefined
    const options: any = {}
    if (isRequired) options.required = true
    for (const prop of modelOptions?.properties || []) {
      if (
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        ['validator', 'type', 'required'].includes(prop.key.name)
      ) {
        options[prop.key.name] = s.sliceNode(prop.value)
      }
    }
    const propName =
      expression.arguments[0]?.type === 'StringLiteral'
        ? expression.arguments[0].value
        : 'modelValue'
    props[propName] = Object.keys(options).length
      ? `{ ${Object.entries(options)
          .map(([key, value]) => `${key}: ${value}`)
          .join(', ')} }`
      : null
    props[`onUpdate:${propName}`] = null
    props[`${propName === 'modelValue' ? 'model' : propName}Modifiers`] = null
  }

  const propsString = Object.entries(props)
    .map(([key, value]) => `'${key}': ${value}`)
    .join(', ')
  if (propsString) {
    const argument = map.defineComponent.arguments[1]
    if (!argument) {
      s.appendRight(
        root.end!,
        `, {${hasRestProp ? 'inheritAttrs: false,' : ''} props: { ${propsString} } }`,
      )
    } else if (argument.type === 'ObjectExpression') {
      prependObjectExpression(argument, 'props', `{ ${propsString} }`, s)
      if (hasRestProp) {
        prependObjectExpression(argument, 'inheritAttrs', 'false', s)
      }
    }
  }

  transformAwait(root, s)
  transformReturn(root, s, options.lib)
}

function prependObjectExpression(
  argument: ObjectExpression,
  name: string,
  value: string,
  s: MagicStringAST,
) {
  if (
    !argument.properties?.find(
      (prop) =>
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        prop.key.name === name,
    )
  ) {
    s.appendRight(argument.start! + 1, `${name}: ${value},`)
  }
}

function getWalkedIds(root: FunctionalNode, propsName: string) {
  const walkedIds = new Set<string>()
  walkIdentifiers(
    root.body,
    (id, parent) => {
      if (
        id.name === propsName &&
        (parent?.type === 'MemberExpression' ||
          parent?.type === 'OptionalMemberExpression')
      ) {
        const prop =
          parent.property.type === 'Identifier'
            ? parent.property.name
            : parent.property.type === 'StringLiteral'
              ? parent.property.value
              : ''
        if (prop) walkedIds.add(prop)
      }
    },
    false,
  )
  return walkedIds
}
