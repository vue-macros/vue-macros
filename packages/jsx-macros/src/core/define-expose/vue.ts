import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import { useExposeHelperId } from '../helper'
import type { CallExpression } from '@babel/types'

export function transformVueDefineExpose(
  node: CallExpression,
  s: MagicStringAST,
  lib: string,
): void {
  s.overwriteNode(
    node.callee,
    importHelperFn(s, 0, 'useExpose', undefined, useExposeHelperId),
  )
  s.appendRight(
    node.arguments[0]?.start || node.end! - 1,
    lib.includes('vapor')
      ? `${importHelperFn(s, 0, 'currentInstance')}, `
      : `${importHelperFn(s, 0, 'getCurrentInstance')}(), `,
  )
}
