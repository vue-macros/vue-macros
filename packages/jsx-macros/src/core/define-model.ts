import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import { useModelHelperId } from './helper'
import type { CallExpression } from '@babel/types'

export function transformDefineModel(
  node: CallExpression,
  propsName: string,
  s: MagicStringAST,
): void {
  s.overwriteNode(
    node.callee,
    importHelperFn(s, 0, 'useModel', undefined, useModelHelperId),
  )
  s.appendRight(
    node.arguments[0]?.start || node.end! - 1,
    `${propsName}, ${
      node.arguments[0]?.type !== 'StringLiteral' ? `'modelValue',` : ''
    }`,
  )
}
