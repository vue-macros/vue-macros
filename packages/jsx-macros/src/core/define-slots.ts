import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import type { CallExpression } from '@babel/types'

export function transformDefineSlots(
  node: CallExpression,
  propsName: string,
  s: MagicStringAST,
  lib: string,
): void {
  s.overwrite(
    node.start!,
    (node.arguments[0]?.start && node.arguments[0].start - 1) ||
      node.typeArguments?.end ||
      node.callee.end!,
    `Object.assign`,
  )
  const slots = lib.includes('vue')
    ? `${importHelperFn(s, 0, 'useSlots')}()`
    : `${propsName}.vSlots`
  s.appendLeft(node.end! - 1, `${node.arguments[0] ? ',' : '{}, '}${slots}`)
}
