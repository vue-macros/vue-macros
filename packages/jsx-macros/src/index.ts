import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { plugin, type Options, type OptionsResolved } from './core/plugin'

export type { Options, OptionsResolved }

const unplugin: UnpluginInstance<Options | undefined, false> =
  createUnplugin(plugin)
export default unplugin
