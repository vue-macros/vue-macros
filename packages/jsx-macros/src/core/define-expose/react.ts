import {
  HELPER_PREFIX,
  importHelperFn,
  type MagicStringAST,
} from '@vue-macros/common'
import { walkIdentifiers } from 'vue/compiler-sfc'
import { getParamsStart, type FunctionalNode } from '..'
import type { CallExpression, Node } from '@babel/types'

export function transformReactDefineExpose(
  node: CallExpression,
  propsName: string,
  root: FunctionalNode,
  s: MagicStringAST,
  lib: string,
  version: number,
): void {
  const useImperativeHandle = importHelperFn(
    s,
    0,
    'useImperativeHandle',
    undefined,
    lib === 'preact' ? 'preact/hooks' : lib,
  )
  const isReact19 = lib === 'react' && version >= 19
  let refName = ''
  if (isReact19) {
    if (root.params[0]?.type === 'ObjectPattern') {
      for (const prop of root.params[0].properties) {
        if (
          prop.type === 'ObjectProperty' &&
          prop.key.type === 'Identifier' &&
          prop.key.name === 'ref'
        ) {
          refName =
            prop.value.type === 'Identifier' ? prop.value.name : prop.key.name
          break
        }
      }
    } else {
      refName = `${propsName}.ref`
    }
  } else {
    const forwardRef = importHelperFn(
      s,
      0,
      'forwardRef',
      undefined,
      lib === 'preact' ? 'preact/compat' : lib,
    )
    if (root.type === 'FunctionDeclaration' && root.id) {
      s.appendLeft(root.start!, `const ${root.id.name} = `)
    }
    s.appendLeft(root.start!, `${forwardRef}(`)

    refName = root.params[1]
      ? s.sliceNode(root.params[1])
      : `${HELPER_PREFIX}ref`
    if (!root.params[0]) {
      s.appendRight(getParamsStart(root, s.original), `, ${refName}`)
    } else if (!root.params[1]) {
      s.appendLeft(root.params[0].end!, `, ${refName}`)
    }
    s.appendRight(root.end!, `)`)
  }

  s.overwrite(
    node.start!,
    node.arguments[0].start!,
    `${useImperativeHandle}(${refName}, () =>(`,
  )
  const result = new Set()
  walkIdentifiers(
    node.arguments[0],
    (id, _, parentStack, ___, isLocal) => {
      if (!isLocal) {
        let res: Node | null = id
        for (let i = parentStack.length - 1; i >= 0; i--) {
          if (
            ['MemberExpression', 'OptionalMemberExpression'].includes(
              parentStack[i].type,
            )
          ) {
            res = parentStack[i]
          } else {
            if (
              ['CallExpression', 'OptionalCallExpression'].includes(
                parentStack[i].type,
              )
            ) {
              res = null
            }
            break
          }
        }
        if (res) result.add(s.sliceNode(res))
      }
    },
    false,
  )
  s.appendRight(
    node.arguments[0].end!,
    `), ${`[${result.size ? [...result] : ''}]`}`,
  )
}
