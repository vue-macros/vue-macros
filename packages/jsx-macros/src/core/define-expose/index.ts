import type { FunctionalNode } from '..'
import { transformReactDefineExpose } from './react'
import { transformVueDefineExpose } from './vue'
import type { CallExpression } from '@babel/types'
import type { MagicStringAST } from '@vue-macros/common'

export { transformVueDefineExpose } from './vue'
export { transformReactDefineExpose } from './react'

export function transformDefineExpose(
  node: CallExpression,
  propsName: string,
  root: FunctionalNode,
  s: MagicStringAST,
  lib: string,
  version: number,
): void {
  if (lib.includes('vue')) {
    transformVueDefineExpose(node, s, lib)
  } else if (lib.includes('react')) {
    transformReactDefineExpose(node, propsName, root, s, lib, version)
  }
}
