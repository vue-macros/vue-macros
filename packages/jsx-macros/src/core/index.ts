import {
  babelParse,
  generateTransform,
  getLang,
  HELPER_PREFIX,
  MagicStringAST,
  walkAST,
  type CodeTransform,
} from '@vue-macros/common'
import type { OptionsResolved } from '..'
import { transformDefineComponent } from './define-component'
import { transformDefineExpose } from './define-expose'
import { transformDefineModel } from './define-model'
import { transformDefineSlots } from './define-slots'
import { transformDefineStyle } from './define-style'
import type {
  ArrowFunctionExpression,
  CallExpression,
  FunctionDeclaration,
  FunctionExpression,
  Node,
  Program,
} from '@babel/types'

export { restructure } from './restructure'

export type FunctionalNode =
  | FunctionDeclaration
  | FunctionExpression
  | ArrowFunctionExpression

export type DefineStyle = {
  expression: CallExpression
  isDeclaration: boolean
  lang: string
}

export type RootMapValue = {
  defineComponent?: CallExpression
  defineModel?: {
    expression: CallExpression
    isRequired: boolean
  }[]
  defineSlots?: CallExpression
  defineExpose?: CallExpression
  defineStyle?: DefineStyle[]
}

export function transformJsxMacros(
  code: string,
  id: string,
  importMap: Map<string, string>,
  options: OptionsResolved,
): CodeTransform | undefined {
  const s = new MagicStringAST(code)
  const ast = babelParse(s.original, getLang(id))
  const rootMap = getRootMap(ast, s, options)

  let defineStyleIndex = 0
  for (const [root, map] of rootMap) {
    map.defineStyle?.forEach((defineStyle) => {
      transformDefineStyle(defineStyle, defineStyleIndex++, root, s, importMap)
    })

    if (root === undefined) continue

    let propsName = `${HELPER_PREFIX}props`
    if (root.params[0]) {
      if (root.params[0].type === 'Identifier') {
        propsName = root.params[0].name
      } else if (root.params[0].type === 'ObjectPattern') {
        const lastProp = root.params[0].properties.at(-1)
        if (
          !map.defineComponent &&
          lastProp?.type === 'RestElement' &&
          lastProp.argument.type === 'Identifier'
        ) {
          propsName = lastProp.argument.name
        } else {
          s.appendRight(
            root.params[0].extra?.trailingComma
              ? (root.params[0].extra?.trailingComma as number) + 1
              : lastProp?.end || root.params[0].end! - 1,
            `${
              !root.params[0].extra?.trailingComma &&
              root.params[0].properties.length
                ? ','
                : ''
            } ...${HELPER_PREFIX}props`,
          )
        }
      }
    } else {
      s.appendRight(getParamsStart(root, s.original), propsName)
    }

    if (map.defineComponent) {
      transformDefineComponent(root, propsName, map, s, ast, options)
    }
    if (map.defineModel?.length) {
      map.defineModel.forEach(({ expression }) => {
        transformDefineModel(expression, propsName, s)
      })
    }
    if (map.defineSlots) {
      transformDefineSlots(map.defineSlots, propsName, s, options.lib)
    }
    if (map.defineExpose) {
      transformDefineExpose(
        map.defineExpose,
        propsName,
        root,
        s,
        options.lib,
        options.version,
      )
    }
  }

  return generateTransform(s, id)
}

function getRootMap(ast: Program, s: MagicStringAST, options: OptionsResolved) {
  const parents: (Node | undefined | null)[] = []
  const rootMap = new Map<FunctionalNode | undefined, RootMapValue>()
  walkAST<Node>(ast, {
    enter(node, parent) {
      parents.unshift(parent)
      const root = isFunctionalNode(parents[1]) ? parents[1] : undefined

      if (
        root &&
        parents[2]?.type === 'CallExpression' &&
        options.defineComponent.alias.includes(s.sliceNode(parents[2].callee))
      ) {
        if (!rootMap.has(root)) rootMap.set(root, {})
        if (!rootMap.get(root)!.defineComponent) {
          rootMap.get(root)!.defineComponent = parents[2]
        }
      }

      const expression =
        node.type === 'VariableDeclaration'
          ? node.declarations[0].init?.type === 'CallExpression' &&
            s.sliceNode(node.declarations[0].init.callee) === '$'
            ? node.declarations[0].init.arguments[0]
            : node.declarations[0].init
          : node.type === 'ExpressionStatement'
            ? node.expression
            : undefined
      if (!expression) return
      const macroExpression = getMacroExpression(expression, options)
      if (!macroExpression) return
      if (!rootMap.has(root)) rootMap.set(root, {})
      const macroName = s.sliceNode(
        macroExpression.callee.type === 'MemberExpression'
          ? macroExpression.callee.object
          : macroExpression.callee,
      )
      if (macroName) {
        if (options.defineModel.alias.includes(macroName)) {
          ;(rootMap.get(root)!.defineModel ??= []).push({
            expression: macroExpression,
            isRequired: expression.type === 'TSNonNullExpression',
          })
        } else if (options.defineStyle.alias.includes(macroName)) {
          const lang =
            macroExpression.callee.type === 'MemberExpression' &&
            macroExpression.callee.property.type === 'Identifier'
              ? macroExpression.callee.property.name
              : 'css'
          ;(rootMap.get(root)!.defineStyle ??= []).push({
            expression: macroExpression,
            isDeclaration: node.type === 'VariableDeclaration',
            lang,
          })
        } else if (options.defineSlots.alias.includes(macroName)) {
          rootMap.get(root)!.defineSlots = macroExpression
        } else if (options.defineExpose.alias.includes(macroName)) {
          rootMap.get(root)!.defineExpose = macroExpression
        }
      }
    },
    leave() {
      parents.shift()
    },
  })
  return rootMap
}

export function isFunctionalNode(node?: Node | null): node is FunctionalNode {
  return !!(
    node &&
    (node.type === 'ArrowFunctionExpression' ||
      node.type === 'FunctionDeclaration' ||
      node.type === 'FunctionExpression')
  )
}

export function getMacroExpression(
  node: Node,
  options: OptionsResolved,
): CallExpression | undefined {
  if (node.type === 'TSNonNullExpression') {
    node = node.expression
  }

  if (node.type === 'CallExpression') {
    if (
      node.callee.type === 'MemberExpression' &&
      node.callee.object.type === 'Identifier' &&
      node.callee.object.name === 'defineStyle'
    ) {
      return node
    } else if (
      node.callee.type === 'Identifier' &&
      [
        ...options.defineComponent.alias,
        ...options.defineSlots.alias,
        ...options.defineModel.alias,
        ...options.defineExpose.alias,
        ...options.defineStyle.alias,
      ].includes(node.callee.name!)
    ) {
      return node
    }
  }
}

export function getParamsStart(node: FunctionalNode, code: string): number {
  return node.params[0]
    ? node.params[0].start!
    : node.start! +
        (code.slice(node.start!, node.body.start!).match(/\(\s*\)/)?.index ||
          0) +
        1
}
