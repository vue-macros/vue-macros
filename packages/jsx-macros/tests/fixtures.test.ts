import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxMacros } from '../src/core'

const options = {
  defineModel: { alias: ['defineModel'] },
  defineSlots: { alias: ['defineSlots'] },
  defineStyle: { alias: ['defineStyle'] },
  defineExpose: { alias: ['defineExpose'] },
  defineComponent: { alias: ['defineComponent'] },
}

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/*.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'vue',
        include: ['*.tsx'],
        version: 3.5,
        ...options,
      })?.code,
  )
})

describe('vue/vapor fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/*.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'vue/vapor',
        include: ['*.tsx'],
        version: 3.5,
        ...options,
      })?.code,
  )
})

describe('react fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/define-expose.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'react',
        include: ['*.tsx'],
        version: 18,
        ...options,
      })?.code,
  )
})

describe('react19 fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/define-expose.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'react',
        include: ['*.tsx'],
        version: 19,
        ...options,
      })?.code,
  )
})
