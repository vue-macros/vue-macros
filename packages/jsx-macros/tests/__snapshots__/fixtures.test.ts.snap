// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/define-component.tsx 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "/vue-macros/jsx-macros/with-defaults";
import { useAttrs as __MACROS_useAttrs } from "vue";
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";
import { withAsyncContext as __MACROS_withAsyncContext } from "vue";import { defineComponent, nextTick } from 'vue'

const Comp = defineComponent(
  (__MACROS_props) => {
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.bar': 'bar'!});const attrs = __MACROS_useAttrs();
    const foo = $(
      __MACROS_useModel(__MACROS_props, 'foo', {
        validator: (value) => {
          return value === 'foo'
        },
        type: String,
      })!,
    )
    return () => <div>{[foo, __MACROS_default_props.bar, attrs.baz]}</div>
  },
  {props: { 'bar': { required: true }, 'foo': { required: true, validator: (value) => {
          return value === 'foo'
        }, type: String }, 'onUpdate:foo': null, 'fooModifiers': null },inheritAttrs: false, name: 'Comp' },
)

const Comp1 = defineComponent((props: { bar: 'bar'; 'onUpdate:bar': any }) => {
  const foo = __MACROS_useModel(props, 'foo')
  return () => <div>{[foo.value, props['bar'], props['onUpdate:bar']]}</div>
}, { props: { 'bar': null, 'onUpdate:bar': null, 'foo': null, 'onUpdate:foo': null, 'fooModifiers': null } })

const Comp2 = defineComponent(async (__MACROS_props) => {
let __temp, __restore

  ;(
  ([__temp,__restore] = __MACROS_withAsyncContext(() => nextTick())),
  await __temp,
  __restore()
)
  let foo = (
  ([__temp,__restore] = __MACROS_withAsyncContext(() => new Promise((resolve) => {
    setTimeout(() => resolve('foo'), 1000)
  }))),
  __temp = await __temp,
  __restore(),
  __temp
)
  return () => <div>{foo}</div>
})
"
`;

exports[`fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { useExpose as __MACROS_useExpose } from "/vue-macros/jsx-macros/use-expose";
import { getCurrentInstance as __MACROS_getCurrentInstance } from "vue";export function Comp(__MACROS_props) {
  __MACROS_useExpose(__MACROS_getCurrentInstance(), {
    foo: 1,
  })
  return <div />
}

export const Comp1 = function (props: any) {
  __MACROS_useExpose(__MACROS_getCurrentInstance(), {
    foo: props.foo,
  })
  return <div />
}

export const Comp2 = ({ foo, ...__MACROS_props }: any) => {
  __MACROS_useExpose(__MACROS_getCurrentInstance(), {
    foo,
  })
  return <div />
}
"
`;

exports[`fixtures > ./fixtures/define-model.tsx 1`] = `
"
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";export const Comp = ({ bar, ...__MACROS_props }: { bar: string }) => {
  const foo = __MACROS_useModel(__MACROS_props, 'foo', { default: bar })
  return <div>{foo.value}</div>
}

export default function (__MACROS_props) {
  const modelValue = $(__MACROS_useModel<string>(__MACROS_props, 'modelValue',)!)
  return (
    <Comp v-model:foo={modelValue} bar="bar">
      {modelValue}
    </Comp>
  )
}
"
`;

exports[`fixtures > ./fixtures/define-slots.tsx 1`] = `
"
import { useSlots as __MACROS_useSlots } from "vue";export const Comp = (__MACROS_props) => {
  const slots = Object.assign<{
    default: () => any
  }>({}, __MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}

export default function (__MACROS_props) {
  const slots = Object.assign({
    default: () => <div>default</div>,
  },__MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}
"
`;

exports[`fixtures > ./fixtures/define-style.tsx 1`] = `
"import "/vue-macros/jsx-macros/define-style/0?scopeId=4e9d5cd0&scoped=true&lang.css";import style1 from "/vue-macros/jsx-macros/define-style/1?scopeId=426a859d&scoped=true&lang.module.scss";import "/vue-macros/jsx-macros/define-style/2?scopeId=63a7910c&scoped=false&lang.scss";import { defineComponent, ref } from 'vue'

export const Comp = (__MACROS_props) => {
  const color = ref('red')
  
  return <div {...{style:{'--4e9d5cd0-color-value': color.value}}} data-v-4e9d5cd0="" class="foo">foo</div>
}

export default defineComponent((__MACROS_props) => {
  const color = ref('red')
  const styles = style1
  return () => (
    <>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class={styles.bar}>foo</div>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class="bar">
        <span data-v-426a859d="">bar</span>
      </div>
    </>
  )
})


"
`;

exports[`react fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { useImperativeHandle as __MACROS_useImperativeHandle } from "react";
import { forwardRef as __MACROS_forwardRef } from "react";export const Comp = __MACROS_forwardRef(function Comp(__MACROS_props, __MACROS_ref) {
  __MACROS_useImperativeHandle(__MACROS_ref, () =>({
    foo: 1,
  }), [])
  return <div />
})

export const Comp1 = __MACROS_forwardRef(function (props: any, __MACROS_ref) {
  __MACROS_useImperativeHandle(__MACROS_ref, () =>({
    foo: props.foo,
  }), [props.foo])
  return <div />
})

export const Comp2 = __MACROS_forwardRef(({ foo, ...__MACROS_props }: any, __MACROS_ref) => {
  __MACROS_useImperativeHandle(__MACROS_ref, () =>({
    foo,
  }), [foo])
  return <div />
})
"
`;

exports[`react19 fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { useImperativeHandle as __MACROS_useImperativeHandle } from "react";export function Comp(__MACROS_props) {
  __MACROS_useImperativeHandle(__MACROS_props.ref, () =>({
    foo: 1,
  }), [])
  return <div />
}

export const Comp1 = function (props: any) {
  __MACROS_useImperativeHandle(props.ref, () =>({
    foo: props.foo,
  }), [props.foo])
  return <div />
}

export const Comp2 = ({ foo, ...__MACROS_props }: any) => {
  __MACROS_useImperativeHandle(, () =>({
    foo,
  }), [foo])
  return <div />
}
"
`;

exports[`vue/vapor fixtures > ./fixtures/define-component.tsx 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "/vue-macros/jsx-macros/with-defaults";
import { useAttrs as __MACROS_useAttrs } from "vue";
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";
import { withAsyncContext as __MACROS_withAsyncContext } from "vue";import { defineComponent, nextTick } from 'vue'

const Comp = defineComponent(
  (__MACROS_props) => {
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.bar': 'bar'!});const attrs = __MACROS_useAttrs();
    const foo = $(
      __MACROS_useModel(__MACROS_props, 'foo', {
        validator: (value) => {
          return value === 'foo'
        },
        type: String,
      })!,
    )
    return <div>{[foo, __MACROS_default_props.bar, attrs.baz]}</div>
  },
  {props: { 'bar': { required: true }, 'foo': { required: true, validator: (value) => {
          return value === 'foo'
        }, type: String }, 'onUpdate:foo': null, 'fooModifiers': null },inheritAttrs: false, name: 'Comp' },
)

const Comp1 = defineComponent((props: { bar: 'bar'; 'onUpdate:bar': any }) => {
  const foo = __MACROS_useModel(props, 'foo')
  return <div>{[foo.value, props['bar'], props['onUpdate:bar']]}</div>
}, { props: { 'bar': null, 'onUpdate:bar': null, 'foo': null, 'onUpdate:foo': null, 'fooModifiers': null } })

const Comp2 = defineComponent(async (__MACROS_props) => {
let __temp, __restore

  ;(
  ([__temp,__restore] = __MACROS_withAsyncContext(() => nextTick())),
  await __temp,
  __restore()
)
  let foo = (
  ([__temp,__restore] = __MACROS_withAsyncContext(() => new Promise((resolve) => {
    setTimeout(() => resolve('foo'), 1000)
  }))),
  __temp = await __temp,
  __restore(),
  __temp
)
  return () => <div>{foo}</div>
})
"
`;

exports[`vue/vapor fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { useExpose as __MACROS_useExpose } from "/vue-macros/jsx-macros/use-expose";
import { currentInstance as __MACROS_currentInstance } from "vue";export function Comp(__MACROS_props) {
  __MACROS_useExpose(__MACROS_currentInstance, {
    foo: 1,
  })
  return <div />
}

export const Comp1 = function (props: any) {
  __MACROS_useExpose(__MACROS_currentInstance, {
    foo: props.foo,
  })
  return <div />
}

export const Comp2 = ({ foo, ...__MACROS_props }: any) => {
  __MACROS_useExpose(__MACROS_currentInstance, {
    foo,
  })
  return <div />
}
"
`;

exports[`vue/vapor fixtures > ./fixtures/define-model.tsx 1`] = `
"
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";export const Comp = ({ bar, ...__MACROS_props }: { bar: string }) => {
  const foo = __MACROS_useModel(__MACROS_props, 'foo', { default: bar })
  return <div>{foo.value}</div>
}

export default function (__MACROS_props) {
  const modelValue = $(__MACROS_useModel<string>(__MACROS_props, 'modelValue',)!)
  return (
    <Comp v-model:foo={modelValue} bar="bar">
      {modelValue}
    </Comp>
  )
}
"
`;

exports[`vue/vapor fixtures > ./fixtures/define-slots.tsx 1`] = `
"
import { useSlots as __MACROS_useSlots } from "vue";export const Comp = (__MACROS_props) => {
  const slots = Object.assign<{
    default: () => any
  }>({}, __MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}

export default function (__MACROS_props) {
  const slots = Object.assign({
    default: () => <div>default</div>,
  },__MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}
"
`;

exports[`vue/vapor fixtures > ./fixtures/define-style.tsx 1`] = `
"import "/vue-macros/jsx-macros/define-style/0?scopeId=4e9d5cd0&scoped=true&lang.css";import style1 from "/vue-macros/jsx-macros/define-style/1?scopeId=426a859d&scoped=true&lang.module.scss";import "/vue-macros/jsx-macros/define-style/2?scopeId=63a7910c&scoped=false&lang.scss";import { defineComponent, ref } from 'vue'

export const Comp = (__MACROS_props) => {
  const color = ref('red')
  
  return <div {...{style:{'--4e9d5cd0-color-value': color.value}}} data-v-4e9d5cd0="" class="foo">foo</div>
}

export default defineComponent((__MACROS_props) => {
  const color = ref('red')
  const styles = style1
  return () => (
    <>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class={styles.bar}>foo</div>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class="bar">
        <span data-v-426a859d="">bar</span>
      </div>
    </>
  )
})


"
`;
